### ISSUE TITLE
JWT signing key stuck in browser cache when rotated via Okta API

### ISSUE THREAD

**Issue by jridge** [okta/okta-auth-js#797]
Got a request to file a **GH issue against the Auth JS SKE** when I posted this issue to the Okta Developers forum regarding a "stuck JWT signing key".   The posted issue is at URL:  https://devforum.okta.com/t/jwt-signing-key-stuck-in-browser-cache-bug/14295

I copy/pasted the entire post here since it provides the needed details to reproduce the issue:  
We are experiencing an interesting challenge when rotating the JWT signing key.
When we manually initiate the JWT signing key rotation using Postman and POST to the following URL:

{{url}}/api/v1/authorizationServers/{{authorizationServerId}}/credentials/lifecycle/keyRotate

The POST works fine and after the “keyRotate” call succeeds, we can use the following GET call and see that the old “ACTIVE” key is now “EXPIRED” and a new “ACTIVE” key is available:

{{url}}/api/v1/authorizationServers/{{authorizationServerId}}/credentials/keys

So far so good, but the problem occurs with the Angular application we are using as a client.

This issue can be easily reproduced using the Angular “okta-hosted-login” sample at: GitHub - okta/samples-js-angular: samples-js-angular

Recreating the problem is easy.

    Login successfully to the Angular application (using either Chrome or Edge on Win10).
    Rotate the signing key via Postman.
    Now the browser(s) are broken.

It isn’t obvious at first, but if you Logout and close the browser (even restart the PC if you want) and try logging in again it fails.

Using the browser Developer Tools and clearing the Local Storage does no good, since the stale signing key come right back with the old value.

Troubleshooting indicates the browsers have the old version of the Okta signing keys cached in the browser and they don’t go away.

To unstick the key you can manually clear out the browser cache (for Chrome use): Settings/Privacy & Security/Clear browsing data/Images and Files. - Clear All.

The question is, how can we make this work without having to manually clear the browsers “Images and Files”?

Without this fix, the browser is unable to successfully log back into the Angular application.

I’m pretty sure I’m not the first person to have this issue.

How can I solve this challenge?

**Comment by bryanapellanes-okta** [comment]
@jridge,
Thanks for reaching out!  I'll bring this to the attention of our Javscript developers and someone will respond here.

Thanks for using Okta!

**Comment by bryanapellanes-okta** [comment]
Internal Ref: OKTA-402180

**Comment by jridge** [comment]
Curious if there has been any progress made on this issue?  We are prepping for Production relatively soon and are hoping for some useful feedback prior to then.  Otherwise we will need to proceed as if JWT Signing Key rotation is not really supported by Okta.  This is a pretty standard configuration (Angular UI, Spring resource server backend) and we really do want to use the rotating keys.  Please let us know how we can successfully do so.  Thanks.

**Comment by shuowu** [comment]
@jridge If I understand the issue correct, the cause looks like after the key rotation, the minted tokens (in your client side) become invalid, and stuck in the browser storage, which lead to the failure for the next login.

Our team will prioritize the issue and release a patch for it.

Meanwhile, can you try [introspect](https://developer.okta.com/docs/reference/api/oidc/#introspect) the tokens in the browser sessionStorage (the default storage for tokens), and clear it if it become invalid as a workaround before the patch release?

**Comment by jridge** [comment]
Clearing the sessionStorage doesn't help.  It appears that when the Angular app does its thing the old stale key is once again copied from the Browser cache.  Clearing the stale signing key from the browser is the challenge.  I'm unaware of any techniques that would allow the Angular application to clear the chrome/edge browsers specific entry in the "images and Files”.  If this was feasible (without using RPA) that would possibly be a usable workaround.   I can tell (after a manual rotation) that the wrong key is being retrieved into sessionStorage.  What we are looking for is a way to clear or force the specific cache entry to expire.   Any suggested approaches (aside from switching to introspection) are appreciated.  Thanks.

**Comment by shuowu** [comment]
@jridge You are right, the stale keys should be the issue. As the keys are acquired from the well-know endpoints, and cached based on `cache` config of the [storageManager](https://github.com/okta/okta-auth-js#storagemanager). And with some digging, looks like currently there is no high level config to control the cache behavior of the well-know response. 

One thing you probably can try is to identify the cache storage for the well-know response (as it's neither localStorage nor sessionStorge in your case, probably check cookie), then clear it and fetch it again to get the new keys. You can fetch it by your own implementation by following logic [here](https://github.com/okta/okta-auth-js/blob/master/lib/oidc/endpoints/well-known.ts#L18).

**Comment by jridge** [comment]
Pretty sure the signing key stuck in the cache is NOT in a cookie.  We can manually fix the problem as follows:  manually clear out the browser cache (for Chrome use): Settings/Privacy & Security/Clear browsing data/Images and Files. - Clear All.    If this is something you guys can fix by releasing a patch that would be much appreciated.   I'm hoping you can come up with a solution that doesn't require manually clearing browsed data/image and files.  Thanks and let us know.

**Comment by shuowu** [comment]
@jridge Have you seen any specific error message when the keys are stale in cache?

**Comment by jridge** [comment]
Yes.  Attempt a login via the Angular Application screen & receive something like the following error:  

AuthSdkError: The key id, paPifw94MwpwDx3Wcs-Xz6mbULZSLAW3AQKrXQjPxYk, was not found in the server's keys

NOTE: the key mentioned in the error is the OLD JWT Signing key.   This is cached somewhere in the browsers cache and I have yet to come up with a way to remove it.  We've currently pinned the signing key, but still hoping for a solution so we can reenable the automatic key rotations.   Thanks for pursuing.
